### ISSUE TITLE
`VK_EXT_shader_object` interaction with `VkPipelineLayout` clarification

### ISSUE THREAD

**Issue by Botondar** [KhronosGroup/Vulkan-Docs#2099]
The [document describing VK_EXT_shader_object](https://github.com/KhronosGroup/Vulkan-Docs/blob/main/proposals/VK_EXT_shader_object.adoc) contains one reference to `VkPipelineLayout`:

> Descriptor set layouts and push constant ranges used by each shader are specified directly (not via a `VkPipelineLayout`), though multiple stages can of course point to the same structures.

It seems unclear however what "specified directly" means in this context, as the extension itself does not define any new commands for setting push constants or binding descriptor sets, and the code examples don't contain this kind of state setup.

My initial assumption was that when `VK_EXT_shader_object` is present and shaders have been bound with `vkCmdBindShadersEXT` applications would be allowed to use the existing `vkCmdPushConstants` and `vkCmdBindDescriptorSets` commands by passing `VK_NULL_HANDLE` to the layout - this, however does not seem to work on the 531.54 Nvidia developer drivers. 

Although this may just be a driver bug, additional clarification might be needed in the documentation that describes how state previously set via `VkPipelineLayout` must be set with this extension.

**Comment by pianofingers** [comment]
Yes, the introductory material and/or the shader object chapter wasn't clear enough, but there is additional detail scattered throughout that can be found by searching the spec for "VkShaderEXT":

> For each set n that is statically used by the VkPipeline bound to the pipeline bind point
> used by this command, or by any of the VkShaderEXT objects bound to stages
> corresponding to the pipeline bind point used by this command, a descriptor set must
> have been bound to n at the same pipeline bind point, with a VkPipelineLayout that is
> compatible for set n, with the VkPipelineLayout or VkDescriptorSetLayout array that was
> used to create the current VkPipeline or VkShaderEXT, as described in Pipeline Layout
> Compatibility

**Comment by pianofingers** [comment]
> Two pipeline layouts are defined to be “compatible for set N” if they were created with identically defined descriptor set layouts for sets zero through N

If I'm reading the spec correctly, a shader _A_ can't be compiled fully in isolation if another shader _B_ bound at the same time uses descriptor numbers lower than shader _A_. You have to know about the layouts of lower-numbered descriptor sets when compiling your shader.

This is your standard: _A consequence of layout compatibility is that when the implementation compiles a pipeline layout and maps pipeline resources to implementation resources, the mechanism for set N should only be a function of sets [0..N]._

We've not gone to the wild west where the ABI and bindings are per-stage, or something. But now we can mix and match shaders, so long as they are interface compatible.

Of course, what's confusing, is are we really meant to be creating pipeline layouts just to bind descriptor sets?

Aside: Instead of ordering sets by frequency, one could imagine alternative approaches (newb here, not saying they're good):

- the shader stage that has the fewest variations gets the lowest numbered sets, to minimize total number of shaders compiled. This isn't likely to be an issue, however.
- the 'base' shader gets the lowest numbered sets, and shaders that build on top of it 'inherit' it and add to the interface. This sounds similar to the other idea, in that a 'base' shader in a hierarchy is going to have the fewest variations

**Comment by pdaniell-nv** [comment]
Hi. Thanks for your feedback on the new VK_EXT_shader_object extension. With the initial release, applications have to unfortunately create a pipeline layout to bind descriptor sets. This was a known limitation and we are working to address this as soon as possible.

**Comment by pdaniell-nv** [comment]
The new [VK_NV_per_stage_descriptor_set](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_NV_per_stage_descriptor_set.html) extension allows applications to no longer need a VkPipelineLayout when using shader objects. An implementation of this should arrive in a future [Vulkan beta driver](https://developer.nvidia.com/vulkan-driver). Going forward I hope we can get a multi-vendor version of this developed and released.
